csrc/ir/nodes.cpp

```diff
@@ -2398,6 +2400,12 @@ std::vector<PolymorphicValue> LoadStoreOp::evaluate(
       at::Tensor in_tensor = inputs[0].as<at::Tensor>();
       at::Tensor out_tensor = in_tensor.permute(computePermutation(
           out_tv->getRootDomain(), out_tv->getRFactorDomain()));
+
+      inferAndValidateAllocationSizesAndStrides(out_tensor, out_tv, ee);
```

Comment:
oh, so we are just totally not using allocDomain at all here. :crying_cat_face: 

So the validation here would just always fail when we have permuted/transformed alloc domain?

Reply:
Generally, I think it does make sense to validate with allocation domain, but my concern are:
1. `inferAndValidateAllocationSizesAndStrides` is slow, and IIUC `LoadStoreOp::evaluate` will become a hot path because it will be executing at runtime. At some point, we might need to worry about CPU overhead here.
2. I don't know how the alias analysis works, but, for example, if I have rFactor `[I1, I2]` and allocation `[I2, I1]`, then `view(-1)` should be invalid, and we should do `reshape(-1)`, which is not an alias. I think the alias analysis should take this into consideration. I think we can not really have an accurate check unless that is built out. But depending on how you would like to incrementally approach, I have no preference on whether we should add this check first or add complete handling allocation first.

Reply:
> inferAndValidateAllocationSizesAndStrides is slow, and IIUC LoadStoreOp::evaluate will become a hot path because it will be executing at runtime. At some point, we might need to worry about CPU overhead here.

Good point -- I'm quite concerned about that too. @jjsjann123 and I happened to talk about this the other day. An alternative is to do more symbolic execution at compile time. For example, if fusion IR could represent strides (as `Val*`s) and propagate these symbolic strides through meta ops like reshapes and transposes, we could spend less run time. This would involve a much larger change (possibly redefine allocation domain) so I didn't take that path for now. 

cc @tfogal

> I think the alias analysis should take this into consideration.

It does, e.g., https://github.com/NVIDIA/Fuser/blob/6751f27a8e128c427f6fd4aedd58732e37fb4013/csrc/optimization/alias_analysis.cpp#L103. The invariant I've been enforcing is: if alias analysis reports `out_tv` is an alias of `in_tv`, then ExpressionEvaluator is able to bind `in_tv` to any valid `in_tensor` (i.e. matching allocation domain and contiguity etc) and evaluate `out_tv` as a `out_tensor` that's an alias of `in_tensor`.

Reply:
> oh, so we are just totally not using allocDomain at all here. 😿
> 
> So the validation here would just always fail when we have permuted/transformed alloc domain?

Right. In fact, I don't even know how to handle an allocation domain that's not merely a permutation of the rfactor domain (see my TODO below). For example, let's say the rfactor domain is [N,H,W,C] and the allocation domain is [N,C/32,H,W,32]. `out_tensor` must still have the shape of [N,H,W,C] (because the rfactor domain dictates the logical shape), but what are going to be the strides?! 

PS: It's interesting that, at this point, fusion IR is more expressive than ATen ops in some cases (e.g. the allocation domain example above) and less expressive in some other cases (e.g. not being able to represent strides). I'm getting nervous about inconsistency like this will bite us at some point in the future.

Reply:
> less expressive in some other cases (e.g. not being able to represent strides)

This feels more like a implementation thing. Up until this point, we only deal with strides in our indexing. So the concept of stride isn't part of our fusion IR.
I'm not saying your concern isn't legit. I don't know if there's something in our indexing code that can be unified/repurposed to represent stride.

Reply:
> I'm getting nervous about inconsistency like this will bite us at some point in the future.

For the place where we need to interface with ATen, we should check consistency and disallow things like `[N, C/32, H, W, 32]` but still allow things like `[N, H, W, C/32, 32]`. For intermediate tensors, I don't think we should have any restriction on what allocation domain can be.

Reply:
> For the place where we need to interface with ATen

Due to the power of `ExpressionEvaluator`, it's too hard to differentiate such places. A user can ask `ExpressionEvaluator` to turn any TensorView (including intermediate ones) into an `at::Tensor`. Furthermore, even if the user asks `ExpressionEvaluator` for a fusion output TensorView, `ExpressionEvaluator` internally computes all dependent `TensorView`s, many of which are intermediate, as `at::Tensor`s. 

AFAICS, for this particular case, it's probably best to proceed with the assumption that allocation must be an rfactor permutation. Because I don't see another good way around. Do you all agree? We should probably have another chat about what's the right assumption between allocation and rfactor in general.

Reply:
> AFAICS, for this particular case, it's probably best to proceed with the assumption that allocation must be an rfactor permutation.

This is definitely better than just ignoring the allocation domain and produce silent wrong result. I think we can start from here.